[RemoveDIs] Load into new debug info format by default in LLVM

SLTozer opened the pull request:
This patch enables parsing and creating modules directly into the new debug info format. Prior to this patch, all modules were constructed with the old debug info format by default, and would be converted into the new format just before running LLVM passes. This is an important milestone, in that this means that every tool will now be exposed to debug records, rather than those that run LLVM passes. As far as I've tested, all LLVM tools/projects now either handle debug records, or convert them to the old intrinsic format.

There are a few unit tests that need updating for this patch; these are either cases of tests that previously needed to set the debug info format to function, or tests that depend on the old debug info format in some way. There should be no visible change in the output of any LLVM tool as a result of this patch, although the likelihood of this patch breaking downstream code means an NFC tag might be a little misleading, if not technically incorrect:

This will probably break some downstream tools that don't already handle debug records. If your downstream code breaks as a result of this change, the simplest fix is to convert the module in question to the old debug format before you process it, using `Module::convertFromNewDbgValues()`. For more information about how to handle debug records or about what has changed, see the migration document:
  https://llvm.org/docs/RemoveDIsDebugInfo.html

SLTozer committed:
[RemoveDIs] Load into new debug format by default in LLVM

OCHyams commented on llvm/unittests/IR/DebugInfoTest.cpp:
```diff
@@ -1077,6 +1073,11 @@ TEST(MetadataTest, DbgVariableRecordConversionRoutines) {
+  // For the purpose of this test, set and un-set the command line option
+  // corresponding to UseNewDbgInfoFormat, but only after parsing, to ensure
+  // that the IR starts off in the old format.
+  UseNewDbgInfoFormat = true;
```
At the end of this test is `UseNewDbgInfoFormat = false`. Shouldn't it get reset to whatever `UseNewDbgInfoFormat` was before this assignment?

OCHyams commented on llvm/unittests/Transforms/Utils/LocalTest.cpp:
```diff
@@ -26,6 +27,27 @@
+// Backup all of the existing settings that may be modified when
+// PreserveInputDbgFormat=true, so that when the test is finished we return them
+// (and the "preserve" setting) to their original values.
+auto TempSettingChange() {
```
I wonder if this should be shared somewhere? Maybe not necessary for this - depends if there's any common unit test header you can use? Then you could use it in the other tests too rather then manually saving and restoring 
`PreserveInputDbgFormat` etc. YMMV. Please apply `static` and the rename mentioned at the other instance of this function either way.

EDIT: This doesn't matter much if we're going to fix up the tests immediately.

OCHyams replied:
Ignore the above, the usage in this file gets removed in #90476 anyway.

OCHyams commented on llvm/unittests/Transforms/Utils/LocalTest.cpp:
```diff
@@ -591,6 +626,11 @@ TEST_F(SalvageDebugInfoTest, RecursiveBlockSimplification) {
 
 TEST(Local, wouldInstructionBeTriviallyDead) {
   LLVMContext Ctx;
+  // FIXME: PreserveInputDbgFormat is set to true because this test has
+  // been written to expect debug intrinsics rather than debug records; use the
+  // intrinsic format until we update the test checks.
```
I'd argue this test doesn't make sense for DbgRecords at all, and the comment should say something along the lines of "TODO: Remove when debug intrinsics are fully removed".

OCHyams commented on llvm/unittests/Transforms/Utils/LocalTest.cpp:
```diff
@@ -1284,6 +1334,11 @@ TEST(Local, ExpressionForConstant) {
   LLVMContext C;
+  // FIXME: PreserveInputDbgFormat is set to true because this test has
+  // been written to expect debug intrinsics rather than debug records; use the
+  // intrinsic format until we update the test checks.
```
I think there's probably value in leaving this test as-is until debug intrinscs are completely removed, as not only does it test that DbgVariableRecord RAUW works, but it gives us the additional coverage that it works after converting from debug intrinsics, which is worth something IMO. Could you adjust this comment to reflect that if you agree?

OCHyams reviewed (commented):
LGTM-ish. Overall looks good, and is an important and exciting milestone indeed!

There's not many of them, but I'm worried about the loss of test coverage between this landing and the tests with FIXMEs getting fixed up. Fixing them up will cause some coverage loss for debug intrinsics, but IMO the primary / default mode should have the better coverage.

I don't think changing these few test alone is enough to suggest that debug intrinsic support is at all unsupported yet, but I think it brings us closer to the point where we need to consider telling people that debug intrinsic support is winding down. We should probably look at the opaque pointer transition - I may be misremembering but IIRC there was a release where typed pointers existed but they had limited testing / support. Ah yep,

```
LLVM 16: Opaque pointers are enabled by default. Typed pointers are supported on a best-effort basis only and not tested.
```
https://llvm.org/docs/OpaquePointers.html

Do you have a plan for those FIXME tests? I'm happy to take a look at fixing those up as a patch to land immediately after this one.

jryans commented:
Should this change (or another one happening soon) mention the new debug info format in the release notes?

OCHyams reviewed (commented):
Once my inline comments are addressed I'm happy to approve this - I've converted the FIXME-tests that can/should be upgraded in #90476, which can land after this does.

dwblaikie commented:
> Should this change (or another one happening soon) mention the new debug info format in the release notes?

+1 to this

SLTozer committed:
Address review comments

SLTozer committed:
Small change to docs link style

SLTozer commented:
Updated some of the tests according to the review comments, except for the comments that are resolved in #90476. Also added some release notes - I added them to the "debug info" section, but it wasn't clear to me whether that refers to debug info in LLVM's IR (superseding the "LLVM IR" category for debug-info-specific topics), or just debug info produced in objects built by LLVM.

SLTozer committed:
Address remaining outstanding comments

OCHyams reviewed (commented):
Code / test changes LGTM

I'll leave it to @dwblaikie or @jryans to Accept if you're happy with the release notes situation.

SLTozer commented on llvm/docs/ReleaseNotes.rst:
```diff
@@ -161,6 +161,12 @@ Changes to the Metadata Info
+  records. This should happen transparently when using the DIBuilder to
+  construct debug variable information, but will require changes for any code
+  that interacts with debug intrinsics directly; for more information, see the
+  `migration docs <https://llvm.org/docs/RemoveDIsDebugInfo.html>`_.
```
That should probably be in the migration docs as well, but it makes sense to put it here.

SLTozer committed:
Update release notes

jryans reviewed (approved):
Looks good to me, thanks for working on this! 😄

SLTozer commented:
Following the failed merge, there's a small functional change and a number of trivial unit test updates needed to fix the patch - since I can't reopen the review, the commit containing the changes is here: https://github.com/SLTozer/llvm-project/commit/a8fed214bfa06d8f06f531f3be505798921a6a93

Anyone able to eyeball it and give reapproval?

jryans commented:
Additional commit looks good to me. 🙂

OCHyams commented:
> Additional commit looks good to me. 🙂

+1

I'd consider splitting up the return condition in `DbgVariableRecord::isKillLocation` into multiple returns, but ymmv, fine either way.

aeubanks commented:
```
Global is external, but doesn't have external or weak linkage!
ptr @_ZN5clang12ast_matchers8internal18makeAllOfCompositeINS_8QualTypeEEENS1_15BindableMatcherIT_EEN4llvm8ArrayRefIPKNS1_7MatcherIS5_EEEE
function declaration may only have a unique !dbg attachment
ptr @_ZN5clang12ast_matchers8internal18makeAllOfCompositeINS_8QualTypeEEENS1_15BindableMatcherIT_EEN4llvm8ArrayRefIPKNS1_7MatcherIS5_EEEE
fatal error: error in backend: Broken module found, compilation aborted!
```
we're seeing a stage 2 ThinLTO (+PGO) build of clang failing with the above verifier error, any ideas what this could be?

OCHyams commented:
The verification errors look very unrelated to DbgRecords, which makes me think that maybe there's some bitcode incompatibility happening here (i.e., new bitcode records being misinterpreted or something).

@aeubanks, this is possibly a long shot, are you able to confirm the version of lld that is used to link the `stage 2` build is the `stage 1` lld?

There was an issue on one of the bots when we first added the new bitcode representation where the `system` lld was used to link both `stage 1` and `stage 2`, which doesn't work because the `stage 1` compiler produced new bitcode that the `system` lld didn't understand: https://github.com/llvm/llvm-zorg/pull/139.

aeubanks commented:
lld is actually not involved here, this is with distributed ThinLTO that invokes clang, and this occurs in the post-link optimization pipeline. I sprinkled some `verifyModule`s around and it seems like [function importing](https://github.com/llvm/llvm-project/blob/8a3277acbc7b7af917f570f7d6430dda41c4d9ee/llvm/lib/LTO/LTOBackend.cpp#L665) is causing the issue. still investigating...

MaskRay commented:
I am sorry but I am going to revert this patch. The stage-2 IR verifier error is from a ThinLTO+SamplePGO build.

`-flto=thin -gsplit-dwarf -g -fdebug-info-for-profiling -fprofile-sample-use=` build of clang, a ThinLTO backend compile has
    assertion failures:

        Global is external, but doesn't have external or weak linkage!
        ptr @_ZN5clang12ast_matchers8internal18makeAllOfCompositeINS_8QualTypeEEENS1_15BindableMatcherIT_EEN4llvm8ArrayRefIPKNS1_7MatcherIS5_EEEE
        function declaration may only have a unique !dbg attachment
        ptr @_ZN5clang12ast_matchers8internal18makeAllOfCompositeINS_8QualTypeEEENS1_15BindableMatcherIT_EEN4llvm8ArrayRefIPKNS1_7MatcherIS5_EEEE

The failures somehow go away if -fprofile-sample-use= is removed.

jmorse commented:
Drive by comment as I'm very out-of-office -- #37394 has a similar error message, during which debug variable locations are implicated in an LTO / IRMover context. The story there is something to do with debug-info materializing something that doesn't have any "real" references, which feels like the kind of thing we might do by mistake with RemoveDIs.

SLTozer commented:
> Another drive-by: in my local build I was able to reliably reproduce unittests failures with this commit for debug build with ` Assertion !NodePtr->isKnownSentinel()' failed.` assertion.

Which unit test(s), and what platform+build configuration were you using?

SLTozer commented:
> I am sorry but I am going to revert this patch. The stage-2 IR verifier error is from a ThinLTO+SamplePGO build.

I've got stage-2 ThinLTO+SPGO build of clang using the flags you included but can't reproduce the error - do you have any more information that might be relevant to reproducing the error? Otherwise I'll continue to dig at the rest of the info you've provided to see if I can see an obvious cause of failure, and see if we can find some way to verify a fix without having to repeatedly land and revert until the fix is confirmed.

danilaml commented:
@SLTozer my issue went away after some time and I wasn't able to reproduce it since. I'm guessing it's some sort of merge fluke.

SLTozer commented:
Ping @MaskRay - any luck so far? And if not, is what I suggested in the post above (submitting just the pre-link bc modules without the sample profile) a possibility? If neither of those, I'd still be interested to know any further details that have turned up - the details so far have narrowed things down a bit, but still haven't gotten me close to figuring out what state the IR would need to be in to trigger this error.

MaskRay commented:
> Ping @MaskRay - any luck so far? And if not, is what I suggested in the post above (submitting just the pre-link bc modules without the sample profile) a possibility? If neither of those, I'd still be interested to know any further details that have turned up - the details so far have narrowed things down a bit, but still haven't gotten me close to figuring out what state the IR would need to be in to trigger this error.

Good news. With this rebased reapply commit (https://github.com/MaskRay/llvm-project/commit/18eaa15964c2857d666fea0409cbb90e8818ee95), I can no longer reproduce the crash ...
Perhaps just reland this...

SLTozer commented:
This certainly is an elusive bug! Hopefully if it shows up again after relanding we'll have better luck identifying the issue.

jeanPerier commented:
I think this patch is causing a buildbot failure with a flang test in the llvm-test-suite using -g and -flto.: https://lab.llvm.org/buildbot/#/builders/198/builds/10772

An assert is hit in llvm::Module::removeDebugIntrinsicDeclarations.

Reproducer:

file.f90:
```
subroutine sub()
 real :: x
end subroutine
```

`flang-new -c -g -flto file.f90`

```
flang-new: llvm/lib/IR/Module.cpp:94: void llvm::Module::removeDebugIntrinsicDeclarations(): Assertion `(!isMaterialized() || DeclareIntrinsicFn->hasZeroLiveUses()) && "Debug declare intrinsic should have had uses removed."' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: bin/flang-new -fc1 -triple x86_64-unknown-linux-gnu -emit-llvm-bc -fcolor-diagnostics -flto=full -mrelocation-model pic -pic-level 2 -pic-is-pie -target-cpu x86-64 -debug-info-kind=standalone -resource-dir bin/.. -mframe-pointer=all -o - -x f95-cpp-input rep_crash.f90
 #0 0x00005635907c86c8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (bin/flang-new+0x2ea86c8)
 #1 0x00005635907c636e llvm::sys::RunSignalHandlers() (bin/flang-new+0x2ea636e)
 #2 0x00005635907c9038 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fcbada42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007fcbada969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007fcbada969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007fcbada969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007fcbada42476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007fcbada287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007fcbada2871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007fcbada39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x0000563594ec414a llvm::Module::removeDebugIntrinsicDeclarations() (bin/flang-new+0x75a414a)
#12 0x0000563593e6ae9b llvm::BitcodeWriterPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (bin/flang-new+0x654ae9b)
#13 0x00005635908f523d llvm::detail::PassModel<llvm::Module, llvm::BitcodeWriterPass, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) FrontendActions.cpp:0:0
#14 0x0000563594ed6356 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (bin/flang-new+0x75b6356)
#15 0x000056359081048d Fortran::frontend::CodeGenAction::runOptimizationPipeline(llvm::raw_pwrite_stream&) (bin/flang-new+0x2ef048d)
#16 0x0000563590812a32 Fortran::frontend::CodeGenAction::executeAction() (bin/flang-new+0x2ef2a32)
#17 0x000056359080349c Fortran::frontend::FrontendAction::execute() (bin/flang-new+0x2ee349c)
#18 0x00005635907df712 Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (bin/flang-new+0x2ebf712)
#19 0x00005635908077ef Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (bin/flang-new+0x2ee77ef)
#20 0x00005635903837d6 fc1_main(llvm::ArrayRef<char const*>, char const*) (bin/flang-new+0x2a637d6)
#21 0x0000563590381e4b main (bin/flang-new+0x2a61e4b)
#22 0x00007fcbada29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#23 0x00007fcbada29e40 call_init ./csu/../csu/libc-start.c:128:20
#24 0x00007fcbada29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#25 0x0000563590381865 _start (bin/flang-new+0x2a61865)
```

I was unable to reproduce when dumping llvm IR and starting back from that, but one can see the llvm-ir with debug info generated for that Fortran program with `flang-new -fc1 -emit-llvm -debug-info-kind=standalone`.

FYI @abidh.

SLTozer commented:
I'll take a look - if it's not a quick fix I'll revert and resubmit with the flang stuff working. To clarify the nature of the error though, LLVM is moving to use a non-intrinsic-based representation of debug info, and this error has occurred because even though the module is set to use debug records, it contains debug intrinsics, which is an error. Since this didn't show up the last time this patch landed, there's presumably some new code added to flang since then that inserts debug intrinsics in some way other than the DIBuilder, so it might be very straightforward to update the code in question.

jeanPerier commented:
Thanks!

> Since this didn't show up the last time this patch landed, there's presumably some new code added to flang since then that inserts debug intrinsics in some way other than the DIBuilder, so it might be very straightforward to update the code in question.

That is very likely, @abidh is currently making patch to generate debug info in flang.

Note that flang is generating MLIR LLVM dialect. It generates mlir::LLVM::DbgDeclareOp [here](https://github.com/llvm/llvm-project/blob/995ba4afcd0d913bca5e082afe90be78b2882f79/flang/lib/Optimizer/CodeGen/CodeGen.cpp#L185C25-L185C49). These MLIR ops are then later translated into LLVM IR (likely in mlir/lib/Target/LLVMIR/ModuleTranslation.cpp or mlir/lib/Target/LLVMIR/DebugTranslation.cpp but I am not sure here).

It is weird though that the flto flag matters here. Is it possible that the -flto pipeline would incorrectly mark the module as using debug records?

SLTozer commented:
Alright, I think I have the fix. Although it's trivial, it's not NFC, and so it should probably be reviewed rather than pushed without commit. I'll revert this for now, and push it up when the other commit lands.
